src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/LiteHashProvider.cs

```diff
@@ -0,0 +1,195 @@
+                return ValueTask.FromCanceled<int>(cancellationToken);
+            }
+
+            LiteHash hash = CreateHash(hashAlgorithmId);
+            return ProcessStreamAsync(hash, source, destination, cancellationToken);
+        }
+
+        internal static ValueTask<byte[]> HashStreamAsync(
+            string hashAlgorithmId,
+            int hashSizeInBytes,
+            Stream source,
+            CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ValueTask.FromCanceled<byte[]>(cancellationToken);
+            }
+
+            LiteHash hash = CreateHash(hashAlgorithmId);
+            return ProcessStreamAsync(hash, source, cancellationToken);
+        }
+
+        internal static int HmacStream(
+            string hashAlgorithmId,
+            int hashSizeInBytes,
+            ReadOnlySpan<byte> key,
+            Stream source,
+            Span<byte> destination)
+        {
+            LiteHmac hash = CreateHmac(hashAlgorithmId, key);
+            return ProcessStream(hash, source, destination);
+        }
+
+        internal static byte[] HmacStream(
+            string hashAlgorithmId,
+            int hashSizeInBytes,
+            ReadOnlySpan<byte> key,
+            Stream source)
+        {
+            byte[] result = new byte[hashSizeInBytes];
+            LiteHmac hash = CreateHmac(hashAlgorithmId, key);
+            int written = ProcessStream(hash, source, result);
+            Debug.Assert(written == hashSizeInBytes);
+            return result;
+        }
+
+        internal static ValueTask<int> HmacStreamAsync(
+            string hashAlgorithmId,
+            int hashSizeInBytes,
+            ReadOnlySpan<byte> key,
+            Stream source,
+            Memory<byte> destination,
+            CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ValueTask.FromCanceled<int>(cancellationToken);
+            }
+
+            LiteHmac hash = CreateHmac(hashAlgorithmId, key);
+            return ProcessStreamAsync(hash, source, destination, cancellationToken);
+        }
+
+        internal static ValueTask<byte[]> HmacStreamAsync(
+            string hashAlgorithmId,
+            int hashSizeInBytes,
+            ReadOnlySpan<byte> key,
+            Stream source,
+            CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ValueTask.FromCanceled<byte[]>(cancellationToken);
+            }
+
+            LiteHmac hash = CreateHmac(hashAlgorithmId, key);
+            return ProcessStreamAsync(hash, source, cancellationToken);
+        }
+
+        /// This takes ownership of the hash parameter and disposes of it when done.
+        private static int ProcessStream<T>(T hash, Stream source, Span<byte> destination) where T : ILiteHash
```

Comment:
I've done something similar (in readonly struct + generic constrained to interface) in a library I contribute to and while it works, any method called on that interface will force a defensive copy, since the compiler is not able to recognize that the struct method itself is readonly. I think it only sees the interface and thus decides on a defensive copy. I'm also not sure anymore whether without the in modifier as you have here, the compiler is able to figure it out properly or not.

Based on the size of the structs used, I assume it's not as relevant here, just wanted to mention it.